duplex_http_call.go

```diff
@@ -113,18 +123,51 @@ func (d *duplexHTTPCall) Send(payload messsagePayload) (int64, error) {
 		// Signal that the stream is closed with the more-typical io.EOF instead of
 		// io.ErrClosedPipe. This makes it easier for protocol-specific wrappers to
 		// match grpc-go's behavior.
-		return bytesWritten, io.EOF
+		err = io.EOF
 	}
 	return bytesWritten, err
 }
 
+func (d *duplexHTTPCall) sendUnary(payload messagePayload) (int64, error) {
+	// Unary messages are sent as a single HTTP request. We don't need to use a
+	// pipe for the request body and we don't need to send headers separately.
+	if !d.requestSent.CompareAndSwap(false, true) {
+		return 0, fmt.Errorf("request already sent")
+	}
+	payloadLength := int64(payload.Len())
+	if payloadLength > 0 {
+		// Build the request body from the payload.
+		payloadBody := newPayloadCloser(payload)
+		d.request.Body = payloadBody
+		d.request.ContentLength = payloadLength
+		d.request.GetBody = func() (io.ReadCloser, error) {
+			if !payloadBody.Rewind() {
+				return nil, fmt.Errorf("payload cannot be retried")
+			}
+			return payloadBody, nil
+		}
+		// Wait on the payloadBody to be completly read or closed before
+		// returning from Send. This ensures that the payload can be reused
+		// after Send returns. See [http.RoundTripper] for more details.
+		defer payloadBody.Wait()
```

Comment:
Actually, even simpler than async release would be to synchronously invalidate the payload here. It could acquire the lock and then clear the underlying message body to nil. That way it can still safely be released, and any concurrent readers will suddenly observe EOF thereafter.

Reply:
> acquire the lock and then clear the underlying message body to nil

The current solution does acquire the lock and then clear the message but waits on Read or Close to signal completion. Reads can still occur after the response is received so invalidating the body would corrupt request messages. The deadlock risk is the same as the streaming case. For streaming a write to the `io.Pipe` will block unless read or closed. For unary a payload will block unless read or closed. The `RoundTripper` implementation must call Close on the request body.

An alternative solution could be to change `bytes.Buffer` for an internal buffer type that we can add reference counts. 

```go
type buffer struct {
    bytes.Buffer
    refCounts atomic.Uint32
}
func (b *buffer) GrabReference() { b.refCounts.Add(1) }
func (b *buffer) ReleaseReference() int32 { b.refCounts.Add(^uint32(-1)) }
```

On fetching a buffer from the pool the refCounts is incremented. On putting a buffer back to the pool the refCount is decremented. This was we can change the buffer lifetime. Wdyt? Could create a new PR for buffer lifetime changes?

Reply:
> The current solution does acquire the lock and then clear the message but waits on Read or Close to signal completion.

Right. I'm saying to **not** wait on Read or Close. Just clear it and let subsequent calls to Read or Close effectively be no-ops, as if the bytes were already fully consumed.

> The deadlock risk is the same as the streaming case.

Ah, that's an interesting point. I wasn't thinking about the blocking nature of writes to the pipe.

But in this case, there's simply no reason to block. At this point, the RPC is done; why add extra non-deterministic latency?

I know I suggested something like ref-counting F2F yesterday, but I think it's actually overkill for this.

Reply:
> Just clear it and let subsequent calls to Read or Close effectively be no-ops

Not waiting is not a valid solution and will corrupt request messages or cause racy behaviour. Reads can and do occur after a response is received. As an example you can mutate the `payloadCloser` to log reads occurring after a wait is called. Here I print the number of bytes read and error for the test: `TestHandlerWithHTTPMaxBytes/grpc/read_max_large`:
```
=== CONT  TestHandlerWithHTTPMaxBytes/grpc/read_max_large
GOT READ after WAIT 16384 <nil>
GOT READ after WAIT 16384 <nil>
GOT READ after WAIT 16384 <nil>
GOT READ after WAIT 16384 <nil>
--- PASS: TestHandlerWithHTTPMaxBytes (0.00s)
    --- PASS: TestHandlerWithHTTPMaxBytes/grpc (0.00s)
        --- PASS: TestHandlerWithHTTPMaxBytes/grpc/read_max_large (0.00s)
PASS
```
This test will keep reading the body but returns an error response.

In almost all other test cases this is difficult to observe or doesn't happen. But the request is always read in a goroutine so not waiting will create racy behaviour. The lock should add minimal overhead as the happy path will be the the single mutex lock and bool check.

Reply:
>  But the request is always read in a goroutine so not waiting will create racy behaviour.

I think we're interpreting the term "racy" in different ways. This thing has a mutex, so the synchronization prevents  any races. Do you mean non-deterministic? As far as the reads after the operation is done, are they important? If they only happen on a "max bytes" issue, where the other side is draining the stream, then transmitting those bytes is just waste; why bother?

If we've already received response headers but we're also still sending request data, that is technically full-duplex operation. It's only even possible with HTTP/2, and I don't see a reason we'd need to support that kind of operation outside of bidirectional streams. You say "corrupt request messages", but in the conditions where this can occur, I think the message that gets corrupted/truncated is necessarily being discarded.

Reply:
> "racy" in different ways

Yes, racy as non-deterministic not data races.

> why bother?

I think there is still value in ensuring the behaviour is maintained. It will correctly handle servers that receive the message concurrently with sending a message. Which I agree should only be in error cases for any frameworks or RPCs I can think of. Saying that if feels more correct to wait than silently discarding the payload. 

The other reason is removing the wait condition has no performance benefit, almost none, under normal requests. The benefit I can see is a slight improvement in readability:
```diff
	p.mu.Lock()
-	for !p.isDone {
-		p.cond.Wait()
-	}
	p.payload = nil
	p.mu.Unlock()
``` 
Would there be other reasons to prefer it?

Reply:
It feels strange to spend any time at all waiting when the RPC is already complete. Also, since this is not a full-duplex operation, it seems inherently risky to block like this. For example, a misbehaving server could fail to exhaust the input until after it's attempted to write a large response, which could lead to a _different_ kind of deadlock due to HTTP/2 flow control windows. In such a case, the server is unable to finish writing the response because the client isn't consuming it, and the client won't consume it because it's waiting on the server to first exhaust its input stream (which the server won't do until after it's finished writing the response, etc).

Reply:
Any misbehaving server/clients can cause a timeout (deadlock) by not progressing the stream. I don't see this as risky behaviour. Probably not an issue for tho so for simplicity have removed the wait in favor of releasing the payload on receiving the response here: 6862f70